component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java

```diff
@@ -858,4 +867,108 @@ private static boolean isURLRelative(String contextFromConfig) throws URISyntaxE
 
         return !new URI(contextFromConfig).isAbsolute();
     }
+
+    /**
+     * Create display name for TOTP in federated Flows.
+     *
+     * @param context Authentication context.
+     * @return TOTP display name for the federated users.
+     * @throws TOTPException When handling identity provider configurations.
+     */
+    public static String createDisplayNameForFederatedUsers(AuthenticationContext context)
+            throws TOTPException {
+
+        SequenceConfig sequenceConfig = context.getSequenceConfig();
+        for (Map.Entry<Integer, StepConfig> entry : sequenceConfig.getStepMap().entrySet()) {
+            StepConfig stepConfig = entry.getValue();
+            AuthenticatorConfig authenticatorConfig = stepConfig.getAuthenticatedAutenticator();
+            ApplicationAuthenticator authenticator = authenticatorConfig.getApplicationAuthenticator();
+
+            if (authenticator instanceof FederatedApplicationAuthenticator) {
+                ExternalIdPConfig externalIdPConfig;
+                String externalIdPConfigName = stepConfig.getAuthenticatedIdP();
+                try {
+                    externalIdPConfig = ConfigurationFacade.getInstance()
+                            .getIdPConfigByName(externalIdPConfigName, context.getTenantDomain());
+                } catch (IdentityProviderManagementException e) {
+                    throw new TOTPException(
+                            "TOTP Display name creation failed!. Error while getting External IDP config.", e);
+                }
+                if (stepConfig != null && stepConfig.isSubjectAttributeStep()) {
+                    if (externalIdPConfig != null) {
+                        Map<String, String> localClaimValues = mapFederateClaimsToLocal(externalIdPConfig, stepConfig, context);
+                        if (localClaimValues.size() == 0 ||
+                                !localClaimValues.containsKey(TOTPAuthenticatorConstants.EMAIL_CLAIM_URL) ||
```

Comment:
If the email is not received, is it possible to consider any other human-readable claim values such as first name, Lastname?